pkg/dataobj/sections/internal/columnar/encoder.go

```diff
@@ -0,0 +1,345 @@
+}
+
+// SetTenant sets the tenant ID for the section. This must be called before
+// calling [Encoder.Flush]. The set tenant is reset after flushing or calling
+// [Encoder.Reset].
+func (enc *Encoder) SetTenant(tenant string) { enc.tenant = tenant }
+
+// SetSortInfo sets the sort order information for the section. This must be
+// called before calling [Encoder.Flush]. The sort order information is reset
+// after flushing or calling [Encoder.Reset].
+func (enc *Encoder) SetSortInfo(info *datasetmd.SortInfo) { enc.sortInfo = info }
+
+// NumColumns returns the number of columns committed to the section.
+func (enc *Encoder) NumColumns() int { return len(enc.columns) }
+
+// OpenColumn opens a new column in the section. OpenColumn fails if there is
+// another open column.
+func (enc *Encoder) OpenColumn(desc *ColumnDesc) (*ColumnEncoder, error) {
+	if enc.curColumn != nil {
+		return nil, errElementExist
+	}
+
+	enc.curColumn = &datasetmd.ColumnDesc{
+		Type: &datasetmd.ColumnType{
+			Physical:   desc.Type.Physical.Proto(),
+			LogicalRef: enc.getDictionaryRef(desc.Type.Logical),
+		},
+		TagRef:           enc.getDictionaryRef(desc.Tag),
+		RowsCount:        uint64(desc.RowsCount),
+		ValuesCount:      uint64(desc.ValuesCount),
+		Compression:      desc.Compression,
+		UncompressedSize: uint64(desc.UncompressedSize),
+		CompressedSize:   uint64(desc.CompressedSize),
+		Statistics:       desc.Statistics,
+
+		// These fields aren't available until the column is closed. We
+		// temporarily set them to the maximum values so they're accounted for
+		// when estimating metadata size.
+
+		PagesCount:           math.MaxUint32,
+		ColumnMetadataOffset: math.MaxUint32,
+		ColumnMetadataLength: math.MaxUint32,
+	}
+
+	return newColumnEncoder(enc, enc.dataSize()), nil
+}
+
+func (enc *Encoder) getDictionaryRef(text string) uint32 {
+	enc.initDictionary() // Make sure the dictionary is initialized.
+
+	if idx, ok := enc.dictionaryLookup[text]; ok {
+		return idx
+	}
+
+	idx := uint32(len(enc.dictionary))
+	enc.dictionary = append(enc.dictionary, text)
+	enc.dictionaryLookup[text] = idx
+
+	return idx
+}
+
+func (enc *Encoder) initDictionary() {
+	if len(enc.dictionary) > 0 && len(enc.dictionaryLookup) > 0 {
+		return // Already initialized.
+	}
+
+	// Initialize the dictionary with index zero being reserved for the
+	// empty string.
+	enc.dictionary = []string{""}
+	enc.dictionaryLookup = map[string]uint32{"": 0}
+}
+
+// dataSize returns the current number of buffered data bytes.
+func (enc *Encoder) dataSize() int {
+	if enc.data == nil {
+		return 0
+	}
+	return enc.data.Len()
+}
+
+// append adds column data and metadata to enc. append must only be called from
+// child elements on Close or Discard. Discard calls must pass nil for both data
+// and metadata to denote a discard.
+//
+// enc *must never* retain data or metadata beyond the call to append.
+func (enc *Encoder) append(pages int, data, metadata []byte) error {
```

Comment:
minor nit: `append` clashes with the keyword, which makes my editor highlight it which is hard for me to scan. If it isn't too much of an issue, could this be named something else? appendData?
I don't mind `append` much, to be clear, so up to you.